---
title: Bioinformatics snippets
author: ~
date: '2017-09-14'
slug: useful-bio
categories:
  - science
tags: []
---



<div id="trim-reads" class="section level2">
<h2>Trim reads</h2>
<pre class="bash"><code>cutadapt -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -o output.fq.gz input.fq.gz</code></pre>
</div>
<div id="map-rna-seq" class="section level2">
<h2>Map RNA-seq</h2>
<pre class="bash"><code># build index
STAR --runThreadN 10 \
     --runMode genomeGenerate \
     --genomeDir genome \
     --genomeFastaFiles genome.fa \
     --sjdbGTFfile genes.gtf
     
# map
STAR --runThreadN 20 \
     --genomeDir genome \
     --alignIntronMax 5000 \
     --alignIntronMin 10 \
     --readFilesCommand zcat \
     --readFilesIn input.fq.gz</code></pre>
</div>
<div id="pseudomap-rna-seq-with-kallisto" class="section level2">
<h2>Pseudomap RNA-seq with kallisto</h2>
<pre class="bash"><code># build index 
kallisto index -i index.idx Arabidopsis_thaliana.TAIR10.cds.all.fa

#quantify
for filename in *.fq.gz; do
    samplename=(${filename//.fq.gz/ })
    kallisto quant -i index.idx -o $samplename -b 100 --single -l 150 -s 20 $filename -t 4
done</code></pre>
</div>
<div id="create-coverage-track-from-bam-file" class="section level2">
<h2>Create coverage track from bam file</h2>
<p><code>bamCoverage</code> is part of <a href="http://deeptools.readthedocs.io/en/latest/index.html">deepTools</a>.</p>
<pre class="bash"><code>bamCoverage -b reads.bam -o coverage.bw</code></pre>
</div>
<div id="take-command-line-arguments" class="section level2">
<h2>Take command-line arguments</h2>
<div id="in-a-shell-script" class="section level3">
<h3>In a shell script</h3>
<p>Required flags are followed by <code>:</code></p>
<pre class="bash"><code>index=  proc=  path=  

while getopts x:pA opt; do
  case $opt in
  x)
      index=$OPTARG
      ;;
  p)
      proc=$OPTARG
      ;;
  A)
      path=${OPTARG%/}
      ;;
  esac
done
shift $((OPTIND - 1))</code></pre>
</div>
<div id="in-a-python-script" class="section level3">
<h3>In a python script</h3>
<pre class="python"><code>from argparse import ArgumentParser

version = pkg_resources.require(&quot;program&quot;)[0].version
parser = ArgumentParser(description=&#39;program description&#39;)
group = parser.add_mutually_exclusive_group()
parser.add_argument(&#39;--version&#39;, action=&#39;version&#39;, version=&#39;%(prog)s &#39;+str(version))
parser.add_argument(&#39;--option&#39;, help=&#39;option description&#39;, required=False, default=False, action=&#39;store_true&#39;)
parser.add_argument(&#39;-n&#39;, &#39;--name&#39;, help=&#39;sample name&#39;, required=True)</code></pre>
</div>
</div>
<div id="gnu-screen" class="section level2">
<h2>GNU Screen</h2>
<p>Start screen: <code>screen -S [screen name]</code><br />
List screens: <code>screen -list</code><br />
Detach: <code>screen -d</code><br />
Attach: <code>screen -r</code><br />
Close screen: <code>ctr-a-d</code><br />
Kill screen: <code>ctr-a :quit</code></p>
</div>
<div id="python-list-comprehension" class="section level2">
<h2>Python list comprehension</h2>
<p>If ever declaring and empty list then using <code>.append</code>, it can be done with list comprehension.</p>
<pre class="python"><code>list3 = []
for x in list2:
    if x[0] in list1:
        list3.append(x)

list3 = [x for x in list2 if x[0] in list1]</code></pre>
</div>
<div id="programming-with-dplyr" class="section level1">
<h1>Programming with dplyr</h1>
<pre class="r"><code>library(dplyr)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<p>See <a href="http://dplyr.tidyverse.org/articles/programming.html" class="uri">http://dplyr.tidyverse.org/articles/programming.html</a></p>
<p>dplyr uses non-standard evaluation which is convienent for interactive use, but makes programming diffucult.</p>
<p>To write functions using dplyr you need to quote the quote the function input using <code>quo()</code>, or quote the input within the function using <code>enquo()</code>. Then, in the dplyr call, unquote the input using <code>!!</code>. Note that this is only possible with dplyr version &gt;= 0.5.0.9. Hereâ€™s an example:</p>
<pre class="r"><code>group_x_summarize &lt;- function(df, x) {
  my_var &lt;- enquo(x)
  df %&gt;% 
    group_by(!!my_var) %&gt;% 
    summarize(count = n())
}</code></pre>
<pre class="r"><code>group_x_summarize(iris, Species)</code></pre>
<pre><code>## # A tibble: 3 x 2
##   Species    count
##   &lt;fct&gt;      &lt;int&gt;
## 1 setosa        50
## 2 versicolor    50
## 3 virginica     50</code></pre>
<pre class="r"><code>group_x_summarize(iris, Sepal.Length) %&gt;% head</code></pre>
<pre><code>## # A tibble: 6 x 2
##   Sepal.Length count
##          &lt;dbl&gt; &lt;int&gt;
## 1          4.3     1
## 2          4.4     3
## 3          4.5     1
## 4          4.6     4
## 5          4.7     2
## 6          4.8     5</code></pre>
<p>If you want to change the output names as well, you need to do this:</p>
<pre class="r"><code>group_x_summarize &lt;- function(df, x, out_name) {
  my_var &lt;- enquo(x)
  my_name &lt;- quo_name(out_name)
  df %&gt;% 
    group_by(!!my_var) %&gt;% 
    summarize(!!my_name := n())
}</code></pre>
<pre class="r"><code>group_x_summarize(iris, Species, &quot;species_count&quot;)</code></pre>
<pre><code>## # A tibble: 3 x 2
##   Species    species_count
##   &lt;fct&gt;              &lt;int&gt;
## 1 setosa                50
## 2 versicolor            50
## 3 virginica             50</code></pre>
<p>Note the use of <code>:=</code> rather than <code>=</code>.</p>
</div>
