<!doctype html>
<html lang="en-us">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title> R demo - Tim Stuart </title>
    <meta name="HandheldFriendly" content="True">
    <meta name="MobileOptimized" content="320">
    <meta name="referrer" content="no-referrer">
    <meta name="description" content="Tim Stuart" />
    <meta property="og:site_name" content="Tim Stuart" />
    <meta property="og:locale" content="nn_NO" />
    <meta property="og:type" content="article" />
    <meta property="og:url" content="/blog/r_demo/" />
    <meta property="og:title" content="R demo" />
    <meta property="og:image" content="/" />
    <meta property="og:description" content="Tim Stuart" />

    
        <meta property="twitter:site" content="@timoast">
    

    <meta property="twitter:title" content="R demo" />
    <meta property="twitter:image" content="/" />
    <meta property="twitter:card" content="summary" />
    <meta property="twitter:description" content="Tim Stuart" />

    <link rel="canonical" href="../../blog/r_demo/">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous" />

    <link rel="stylesheet" href="../../css/custom.css" />

    
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.13.1/styles/tomorrow.min.css" integrity="sha256-0QU8ry64q+N6YBIEF/6XF6vUeF15gbNO4tLS6ikk0FI=" crossorigin="anonymous" />
    

    

    <link rel="shortcut icon"
        href="../../img/favicon.ico">

    
</head>

<body>
    
    <div class="my-4 my-md-5 header">
    <div class="container">
        <div class="row">
            <div class="col-auto offset-md-1 d-none d-md-block">
                
            </div>
            <div class="col-auto align-self-center mr-auto">
                <a href="../../">
                    <h1 class="name">Tim Stuart</h1>
                </a>

                <ul class="nav nav-primary">
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-about" href="../../">
                                
                                About
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-research" href="../../research/">
                                
                                Research
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-publications" href="../../publications/">
                                
                                Publications
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-software" href="../../software/">
                                
                                Software
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-blog" href="../../blog/">
                                
                                Blog
                            </a>
                        </li>
                    

                    
                </ul>

            </div>
        </div>
    </div>
</div>


    <div class="content">
        <div class="container">
            <div class="row justify-content-center">
                <div class="col-sm-12 col-md-10">
                    <h1 class="mx-0 mx-md-4 blog-post-title">R demo</h1>

                    <div class="mb-md-4 meta">
                        
                            
                                <span class="author" title="Tim Stuart">
                                    Tim Stuart
                                </span>
                            
                        

                        <span class="date middot" title='Thu Sep 22 2016 00:00:00 UTC'>
                            2016-09-22
                        </span>

                        <span class="reading-time middot">
                            25 minute read
                        </span>

                        <div class="d-none d-md-inline tags">
                            <ul class="list-unstyled d-inline">
                                
                                    <li class="d-inline middot">
                                        <a href="../../tags/r">R</a>
                                    </li>
                                
                            </ul>
                        </div>

                        <div class="d-none d-md-inline tags">
                            <ul class="list-unstyled d-inline">
                                
                                
                            </ul>
                        </div>
                    </div>

                    <div class="markdown">
                        
    
<script src="../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<script src="../../rmarkdown-libs/plotly-binding/plotly.js"></script>
<script src="../../rmarkdown-libs/typedarray/typedarray.min.js"></script>
<script src="../../rmarkdown-libs/jquery/jquery.min.js"></script>
<link href="../../rmarkdown-libs/crosstalk/css/crosstalk.css" rel="stylesheet" />
<script src="../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<link href="../../rmarkdown-libs/plotly-htmlwidgets-css/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="../../rmarkdown-libs/plotly-main/plotly-latest.min.js"></script>


<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Clone the repo if you haven’t already:</p>
<pre class="bash"><code>git clone https://github.com/timoast/dac.git</code></pre>
<p>Install <a href="https://www.rstudio.com/">RStudio</a>.</p>
<p>Install the following packages:</p>
<pre class="r"><code>install.packages(c(&quot;readr&quot;, &quot;tidyr&quot;, &quot;data.table&quot;, &quot;dplyr&quot;, &quot;ggplot2&quot;))</code></pre>
<pre class="r"><code>library(readr)
library(data.table)
library(ggplot2)
library(tidyr)</code></pre>
<p>All the following code will assume you are in the directory that holds this file (<code>r_demo.rmd</code>) in the github repo (ie <code>dac/code/</code>).</p>
</div>
<div id="importing-data" class="section level2">
<h2>Importing Data</h2>
<div id="the-readr-way" class="section level3">
<h3>The <code>readr</code> way</h3>
<p><a href="https://blog.rstudio.org/2016/08/05/readr-1-0-0/">Some docs</a></p>
<pre class="r"><code>library(readr)

read_tsv(&quot;../data/r_demo/coverages.tsv&quot;, col_names = c(&quot;Accession&quot;, &quot;Coverage&quot;))</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   Accession = col_character(),
##   Coverage = col_double()
## )</code></pre>
<pre><code>## # A tibble: 216 x 2
##    Accession Coverage
##    &lt;chr&gt;        &lt;dbl&gt;
##  1 Aa-0            41
##  2 Abd-0           67
##  3 Ag-0            47
##  4 Ak-1            38
##  5 Alst-1          42
##  6 Altai-5         42
##  7 Amel-1          60
##  8 An-1            39
##  9 Ang-0           48
## 10 Anholt-1        53
## # … with 206 more rows</code></pre>
<pre class="r"><code># also works for gzipped files
read_tsv(&quot;../data/r_demo/mc_data.bed.gz&quot;)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   coords = col_character(),
##   ins0 = col_double(),
##   ins1 = col_double()
## )</code></pre>
<pre><code>## # A tibble: 999 x 3
##    coords              ins0    ins1
##    &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;
##  1 chr1,8668,8746   0       0.00537
##  2 chr1,17024,18924 0.00519 0.00410
##  3 chr1,18331,18642 0.00372 0.00869
##  4 chr1,23022,23145 0       0.00418
##  5 chr1,55676,56576 0.00417 0.0129 
##  6 chr1,69022,69135 0.00823 0.00424
##  7 chr1,69025,69143 0.00348 0.00361
##  8 chr1,69025,69148 0.00278 0.00760
##  9 chr1,69028,69133 0.00549 0      
## 10 chr1,76844,77500 0.00328 0.00379
## # … with 989 more rows</code></pre>
</div>
<div id="the-data.table-way" class="section level3">
<h3>The <code>data.table</code> way</h3>
<p><a href="https://github.com/Rdatatable/data.table/wiki">Some docs</a></p>
<pre class="r"><code>library(data.table)

fread(&quot;../data/r_demo/coverages.tsv&quot;, header = FALSE, col.names = c(&quot;Accession&quot;, &quot;Coverage&quot;))</code></pre>
<pre><code>##      Accession Coverage
##   1:      Aa-0       41
##   2:     Abd-0       67
##   3:      Ag-0       47
##   4:      Ak-1       38
##   5:    Alst-1       42
##  ---                   
## 212:      Ws-2       52
## 213:      Wt-5       29
## 214:      Yo-0       78
## 215:     Zal-1       41
## 216:     Zdr-1       19</code></pre>
<pre class="r"><code>fread(&quot;gzip -dc ../data/r_demo/mc_data.bed.gz&quot;)</code></pre>
<pre><code>##                      coords        ins0        ins1
##   1:         chr1,8668,8746 0.000000000 0.005366726
##   2:       chr1,17024,18924 0.005193408 0.004104893
##   3:       chr1,18331,18642 0.003720434 0.008686003
##   4:       chr1,23022,23145 0.000000000 0.004184100
##   5:       chr1,55676,56576 0.004172405 0.012852597
##  ---                                               
## 995: chr1,12566243,12568303 0.264952130 0.076252087
## 996: chr1,12581911,12582918 0.251092213 0.248537692
## 997: chr1,12584274,12584300 0.082298985 0.096469517
## 998: chr1,12594543,12594591 0.333835719 0.404032186
## 999: chr1,12612939,12613280 0.357346865 0.024760132</code></pre>
<pre class="r"><code>fread(&quot;seq 25 | xargs -n 5 echo&quot;)</code></pre>
<pre><code>##    V1 V2 V3 V4 V5
## 1:  1  2  3  4  5
## 2:  6  7  8  9 10
## 3: 11 12 13 14 15
## 4: 16 17 18 19 20
## 5: 21 22 23 24 25</code></pre>
<pre class="r"><code>fread(&quot;curl https://raw.githubusercontent.com/timoast/Arabidopsis-TE-variants/master/ProcessedData/coverages.tsv&quot;)</code></pre>
<pre><code>##          V1 V2
##   1:   Aa-0 41
##   2:  Abd-0 67
##   3:   Ag-0 47
##   4:   Ak-1 38
##   5: Alst-1 42
##  ---          
## 212:   Ws-2 52
## 213:   Wt-5 29
## 214:   Yo-0 78
## 215:  Zal-1 41
## 216:  Zdr-1 19</code></pre>
<pre class="r"><code>fread(&quot;bedtools intersect -a ../data/r_demo/a.bed -b ../data/r_demo/b.bed&quot;)</code></pre>
<pre><code>##      V1 V2  V3
## 1: chr1 20 100
## 2: chr1 50 100
## 3: chr1 20  50</code></pre>
<p><code>data.table</code> has lots of other useful functions, see <a href="https://www.datacamp.com/courses/data-table-data-manipulation-r-tutorial">datacamp course</a></p>
</div>
<div id="brief-tangent-using-fread-to-eliminate-intermediate-files" class="section level3">
<h3>Brief tangent – using fread to eliminate intermediate files</h3>
<pre class="r"><code>centromere_distance &lt;- fread(&quot;bedtools closest -d -a ../data/r_demo/c_dmrs.tsv.gz -b ../data/r_demo/centromere_positions.txt&quot;)

hist(centromere_distance$V7 / 1000, breaks = 100, col = &quot;blue&quot;,
     main = &quot;C-DMR distances to centromere&quot;,
     xlab = &quot;Distance (kb)&quot;)</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/plotting_example-1.png" width="672" /></p>
<pre class="r"><code>densities &lt;- hist(centromere_distance$V7, breaks = 500, plot = F)
image(as.matrix(densities$counts))</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/plotting_example-2.png" width="672" /></p>
<pre class="r"><code>library(RColorBrewer)
colors &lt;- colorRampPalette(brewer.pal(9,&quot;RdYlBu&quot;))(100)
image(as.matrix(densities$counts), col = rev(colors))</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/plotting_example-3.png" width="672" /></p>
<pre class="r"><code># create a scale bar
image(matrix(seq(100)), col = rev(colors))</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/plotting_example-4.png" width="672" /></p>
</div>
</div>
<div id="tidying-data" class="section level2">
<h2>Tidying data</h2>
<p>Some background: <a href="https://www.jstatsoft.org/article/view/v059i10">Tidy data, Wickham 2014</a></p>
<p>Variables go in the columns, observations go in the rows.</p>
<p>Four main operations in tidying data:</p>
<ul>
<li>Move observations from the column names to rows<br />
</li>
<li>Move variables from rows to columns<br />
</li>
<li>Separate multiple variables into multiple columns<br />
</li>
<li>Gather single variable from many columns into one column</li>
</ul>
<pre class="r"><code>library(tidyr)

data &lt;- read_tsv(&quot;../data/r_demo/experiment.tsv&quot;)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   position = col_double(),
##   experiment_1 = col_double(),
##   experiment_2 = col_double(),
##   experiment_3 = col_double(),
##   experiment_4 = col_double()
## )</code></pre>
<pre class="r"><code>data</code></pre>
<pre><code>## # A tibble: 500 x 5
##    position experiment_1 experiment_2 experiment_3 experiment_4
##       &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
##  1        1        0.295        0           0.691         0.224
##  2        2        0.262        0.194       0.0300        0.282
##  3        3        0.223        0.170       0.647         0.302
##  4        4        0.186        0.353       0.673         0.208
##  5        5        0.253        0.756       0.525         0.284
##  6        6        0.127        0           0.692         0.340
##  7        7        0.148        0.301       0.199         0.401
##  8        8        0.220        0.133       0.319         0.349
##  9        9        0.189        0.474       0.692         0.415
## 10       10        0.197        0.425       0.992         0.380
## # … with 490 more rows</code></pre>
<pre class="r"><code>tidy_data &lt;- gather(data, experiment, mC, experiment_1:experiment_4)
tidy_data</code></pre>
<pre><code>## # A tibble: 2,000 x 3
##    position experiment      mC
##       &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;
##  1        1 experiment_1 0.295
##  2        2 experiment_1 0.262
##  3        3 experiment_1 0.223
##  4        4 experiment_1 0.186
##  5        5 experiment_1 0.253
##  6        6 experiment_1 0.127
##  7        7 experiment_1 0.148
##  8        8 experiment_1 0.220
##  9        9 experiment_1 0.189
## 10       10 experiment_1 0.197
## # … with 1,990 more rows</code></pre>
<pre class="r"><code>spread(tidy_data, experiment, mC)</code></pre>
<pre><code>## # A tibble: 500 x 5
##    position experiment_1 experiment_2 experiment_3 experiment_4
##       &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
##  1        1        0.295        0           0.691         0.224
##  2        2        0.262        0.194       0.0300        0.282
##  3        3        0.223        0.170       0.647         0.302
##  4        4        0.186        0.353       0.673         0.208
##  5        5        0.253        0.756       0.525         0.284
##  6        6        0.127        0           0.692         0.340
##  7        7        0.148        0.301       0.199         0.401
##  8        8        0.220        0.133       0.319         0.349
##  9        9        0.189        0.474       0.692         0.415
## 10       10        0.197        0.425       0.992         0.380
## # … with 490 more rows</code></pre>
<p>Unite and separate columns</p>
<pre class="r"><code>mc &lt;- read_tsv(&quot;../data/r_demo/mc_data.bed.gz&quot;)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   coords = col_character(),
##   ins0 = col_double(),
##   ins1 = col_double()
## )</code></pre>
<pre class="r"><code>head(mc)</code></pre>
<pre><code>## # A tibble: 6 x 3
##   coords              ins0    ins1
##   &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;
## 1 chr1,8668,8746   0       0.00537
## 2 chr1,17024,18924 0.00519 0.00410
## 3 chr1,18331,18642 0.00372 0.00869
## 4 chr1,23022,23145 0       0.00418
## 5 chr1,55676,56576 0.00417 0.0129 
## 6 chr1,69022,69135 0.00823 0.00424</code></pre>
<pre class="r"><code># first column has three variables: chromsome, start, end
new_mc &lt;- separate(mc, coords, c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;))
new_mc</code></pre>
<pre><code>## # A tibble: 999 x 5
##    chr   start end      ins0    ins1
##    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;
##  1 chr1  8668  8746  0       0.00537
##  2 chr1  17024 18924 0.00519 0.00410
##  3 chr1  18331 18642 0.00372 0.00869
##  4 chr1  23022 23145 0       0.00418
##  5 chr1  55676 56576 0.00417 0.0129 
##  6 chr1  69022 69135 0.00823 0.00424
##  7 chr1  69025 69143 0.00348 0.00361
##  8 chr1  69025 69148 0.00278 0.00760
##  9 chr1  69028 69133 0.00549 0      
## 10 chr1  76844 77500 0.00328 0.00379
## # … with 989 more rows</code></pre>
<pre class="r"><code>unite(new_mc, coords, chr, start, end, sep = &quot;,&quot;)</code></pre>
<pre><code>## # A tibble: 999 x 3
##    coords              ins0    ins1
##    &lt;chr&gt;              &lt;dbl&gt;   &lt;dbl&gt;
##  1 chr1,8668,8746   0       0.00537
##  2 chr1,17024,18924 0.00519 0.00410
##  3 chr1,18331,18642 0.00372 0.00869
##  4 chr1,23022,23145 0       0.00418
##  5 chr1,55676,56576 0.00417 0.0129 
##  6 chr1,69022,69135 0.00823 0.00424
##  7 chr1,69025,69143 0.00348 0.00361
##  8 chr1,69025,69148 0.00278 0.00760
##  9 chr1,69028,69133 0.00549 0      
## 10 chr1,76844,77500 0.00328 0.00379
## # … with 989 more rows</code></pre>
</div>
<div id="manipulating-data" class="section level2">
<h2>Manipulating data</h2>
<p>The <a href="https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html">dplyr</a> package:<br />
* <code>filter</code> rows<br />
* <code>select</code> columns<br />
* <code>mutate</code> columns<br />
* <code>arrange</code> rows<br />
* <code>group_by</code> variables</p>
<pre class="r"><code>library(dplyr)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:data.table&#39;:
## 
##     between, first, last</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code># from earlier
coverages &lt;- read_tsv(&quot;../data/r_demo/coverages.tsv&quot;, col_names = c(&quot;Accession&quot;, &quot;Coverage&quot;))</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   Accession = col_character(),
##   Coverage = col_double()
## )</code></pre>
<pre class="r"><code># use mutate to create new columns
cov &lt;- mutate(coverages, high = Coverage &gt; 50)
cov</code></pre>
<pre><code>## # A tibble: 216 x 3
##    Accession Coverage high 
##    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;
##  1 Aa-0            41 FALSE
##  2 Abd-0           67 TRUE 
##  3 Ag-0            47 FALSE
##  4 Ak-1            38 FALSE
##  5 Alst-1          42 FALSE
##  6 Altai-5         42 FALSE
##  7 Amel-1          60 TRUE 
##  8 An-1            39 FALSE
##  9 Ang-0           48 FALSE
## 10 Anholt-1        53 TRUE 
## # … with 206 more rows</code></pre>
<pre class="r"><code># group rows by common variables
cov &lt;- group_by(cov, high)
cov &lt;- mutate(cov, group_average = mean(Coverage))
cov</code></pre>
<pre><code>## # A tibble: 216 x 4
## # Groups:   high [2]
##    Accession Coverage high  group_average
##    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt;
##  1 Aa-0            41 FALSE          37.3
##  2 Abd-0           67 TRUE           70.8
##  3 Ag-0            47 FALSE          37.3
##  4 Ak-1            38 FALSE          37.3
##  5 Alst-1          42 FALSE          37.3
##  6 Altai-5         42 FALSE          37.3
##  7 Amel-1          60 TRUE           70.8
##  8 An-1            39 FALSE          37.3
##  9 Ang-0           48 FALSE          37.3
## 10 Anholt-1        53 TRUE           70.8
## # … with 206 more rows</code></pre>
<pre class="r"><code># use select to choose columns
select(cov, Accession, group_average)</code></pre>
<pre><code>## Adding missing grouping variables: `high`</code></pre>
<pre><code>## # A tibble: 216 x 3
## # Groups:   high [2]
##    high  Accession group_average
##    &lt;lgl&gt; &lt;chr&gt;             &lt;dbl&gt;
##  1 FALSE Aa-0               37.3
##  2 TRUE  Abd-0              70.8
##  3 FALSE Ag-0               37.3
##  4 FALSE Ak-1               37.3
##  5 FALSE Alst-1             37.3
##  6 FALSE Altai-5            37.3
##  7 TRUE  Amel-1             70.8
##  8 FALSE An-1               37.3
##  9 FALSE Ang-0              37.3
## 10 TRUE  Anholt-1           70.8
## # … with 206 more rows</code></pre>
<pre class="r"><code># won&#39;t let you drop a column if the data is grouped by that column
cov &lt;- ungroup(cov)
select(cov, Accession, group_average)</code></pre>
<pre><code>## # A tibble: 216 x 2
##    Accession group_average
##    &lt;chr&gt;             &lt;dbl&gt;
##  1 Aa-0               37.3
##  2 Abd-0              70.8
##  3 Ag-0               37.3
##  4 Ak-1               37.3
##  5 Alst-1             37.3
##  6 Altai-5            37.3
##  7 Amel-1             70.8
##  8 An-1               37.3
##  9 Ang-0              37.3
## 10 Anholt-1           70.8
## # … with 206 more rows</code></pre>
<pre class="r"><code># can alse choose what NOT to include
select(cov, -high)</code></pre>
<pre><code>## # A tibble: 216 x 3
##    Accession Coverage group_average
##    &lt;chr&gt;        &lt;dbl&gt;         &lt;dbl&gt;
##  1 Aa-0            41          37.3
##  2 Abd-0           67          70.8
##  3 Ag-0            47          37.3
##  4 Ak-1            38          37.3
##  5 Alst-1          42          37.3
##  6 Altai-5         42          37.3
##  7 Amel-1          60          70.8
##  8 An-1            39          37.3
##  9 Ang-0           48          37.3
## 10 Anholt-1        53          70.8
## # … with 206 more rows</code></pre>
<pre class="r"><code># sort using arrange
arrange(cov, Coverage)</code></pre>
<pre><code>## # A tibble: 216 x 4
##    Accession     Coverage high  group_average
##    &lt;chr&gt;            &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt;
##  1 Koln                11 FALSE          37.3
##  2 Se-0                15 FALSE          37.3
##  3 Ra-0                17 FALSE          37.3
##  4 Wei-0               17 FALSE          37.3
##  5 Old-1               18 FALSE          37.3
##  6 Pu2-8               19 FALSE          37.3
##  7 Ragl-1              19 FALSE          37.3
##  8 Zdr-1               19 FALSE          37.3
##  9 Bg-2                20 FALSE          37.3
## 10 Buckhorn-Pass       20 FALSE          37.3
## # … with 206 more rows</code></pre>
<pre class="r"><code>arrange(cov, desc(Coverage))</code></pre>
<pre><code>## # A tibble: 216 x 4
##    Accession Coverage high  group_average
##    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt;
##  1 Br-0           133 TRUE           70.8
##  2 Lan-0          127 TRUE           70.8
##  3 Kb-0           110 TRUE           70.8
##  4 El-0           107 TRUE           70.8
##  5 La-0           102 TRUE           70.8
##  6 Kl-5           100 TRUE           70.8
##  7 Kar-1           97 TRUE           70.8
##  8 Gel-1           96 TRUE           70.8
##  9 Nok-3           96 TRUE           70.8
## 10 Rennes-1        96 TRUE           70.8
## # … with 206 more rows</code></pre>
<pre class="r"><code># useful for coordinates
dmrs &lt;- read_tsv(&quot;../data/r_demo/c_dmrs.tsv.gz&quot;, col_names = c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;))</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   chr = col_character(),
##   start = col_double(),
##   end = col_double()
## )</code></pre>
<pre class="r"><code>dmrs</code></pre>
<pre><code>## # A tibble: 13,484 x 3
##    chr    start    end
##    &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;
##  1 chr1     792    911
##  2 chr1   22665  22807
##  3 chr1   41392  41445
##  4 chr1   71457  71476
##  5 chr1   76839  77252
##  6 chr1  193333 193525
##  7 chr1  255844 255893
##  8 chr1  291887 291998
##  9 chr1  355668 356037
## 10 chr1  428072 428073
## # … with 13,474 more rows</code></pre>
<pre class="r"><code>arrange(dmrs, chr, desc(start))</code></pre>
<pre><code>## # A tibble: 13,484 x 3
##    chr      start      end
##    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 chr1  30417725 30417837
##  2 chr1  30416592 30416901
##  3 chr1  30416104 30416138
##  4 chr1  30391058 30391076
##  5 chr1  30390110 30390285
##  6 chr1  30388868 30389254
##  7 chr1  30388263 30388288
##  8 chr1  30376878 30377010
##  9 chr1  30345730 30346096
## 10 chr1  30295026 30295310
## # … with 13,474 more rows</code></pre>
<div id="joining-tables" class="section level3">
<h3>Joining tables</h3>
<p><code>dplyr</code> also has join functions.</p>
<!-- ![](../images/joins.png) -->
<p><img src="../images/joins.png" width="500"></p>
<pre class="r"><code>te_insertion_counts &lt;- read_tsv(&quot;../data/r_demo/insertion_counts.tsv&quot;, col_names = c(&quot;Accession&quot;, &quot;Insertions&quot;))</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   Accession = col_character(),
##   Insertions = col_double()
## )</code></pre>
<pre class="r"><code>left_join(coverages, te_insertion_counts)</code></pre>
<pre><code>## Joining, by = &quot;Accession&quot;</code></pre>
<pre><code>## # A tibble: 216 x 3
##    Accession Coverage Insertions
##    &lt;chr&gt;        &lt;dbl&gt;      &lt;dbl&gt;
##  1 Aa-0            41        367
##  2 Abd-0           67        420
##  3 Ag-0            47        443
##  4 Ak-1            38        305
##  5 Alst-1          42        379
##  6 Altai-5         42        417
##  7 Amel-1          60        330
##  8 An-1            39        313
##  9 Ang-0           48        350
## 10 Anholt-1        53        343
## # … with 206 more rows</code></pre>
</div>
</div>
<div id="put-it-all-together" class="section level2">
<h2>Put it all together</h2>
<p>In unix, the pipe lets you use output from one program/function as input to another. This lets you chain together commands an eliminate intermediate files. It also provides the ability to start one process before the previous has finished. For example you can pipe the output of bowtie into samtools and convert straight to bam without having to wait for the mapping to finish.</p>
<p>The same can be done in R with any function that takes and returns a dataframe. In unix the pipe character is <code>|</code>, in R it is <code>%&gt;%</code>. The pipe is part of the <code>magrittr</code> package, but is also loaded any time you load <code>readr</code>, <code>dplyr</code>, or <code>tidyr</code>.</p>
<pre class="r"><code>df &lt;- read_tsv(&quot;../data/r_demo/coverages.tsv&quot;, col_names = c(&quot;Accession&quot;, &quot;Coverage&quot;)) %&gt;%
  mutate(high = Coverage &gt; 50) %&gt;%
  group_by(high) %&gt;%
  mutate(group_av = mean(Coverage)) %&gt;%
  left_join(te_insertion_counts)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   Accession = col_character(),
##   Coverage = col_double()
## )</code></pre>
<pre><code>## Joining, by = &quot;Accession&quot;</code></pre>
<pre class="r"><code>df</code></pre>
<pre><code>## # A tibble: 216 x 5
## # Groups:   high [2]
##    Accession Coverage high  group_av Insertions
##    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;    &lt;dbl&gt;      &lt;dbl&gt;
##  1 Aa-0            41 FALSE     37.3        367
##  2 Abd-0           67 TRUE      70.8        420
##  3 Ag-0            47 FALSE     37.3        443
##  4 Ak-1            38 FALSE     37.3        305
##  5 Alst-1          42 FALSE     37.3        379
##  6 Altai-5         42 FALSE     37.3        417
##  7 Amel-1          60 TRUE      70.8        330
##  8 An-1            39 FALSE     37.3        313
##  9 Ang-0           48 FALSE     37.3        350
## 10 Anholt-1        53 TRUE      70.8        343
## # … with 206 more rows</code></pre>
</div>
<div id="plotting" class="section level2">
<h2>Plotting</h2>
<pre class="r"><code>library(ggplot2)

ggplot(df, aes(Coverage, Insertions)) +
  geom_point() + geom_smooth(method = &quot;lm&quot;)</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/ggplot-1.png" width="672" /></p>
<pre class="r"><code>ggplot(df, aes(Coverage, Insertions, color = high)) +
  geom_point() + geom_smooth(method = &quot;lm&quot;)</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/ggplot-2.png" width="672" /></p>
<pre class="r"><code>ggplot(df, aes(Coverage, Insertions)) +
  geom_point() + facet_wrap(~high) +
  theme_bw()</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/ggplot-3.png" width="672" /></p>
<pre class="r"><code>ggplot(tidy_data, aes(experiment, mC)) +
  geom_jitter(alpha = 0.5) + theme_bw()</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/ggplot-4.png" width="672" /></p>
<p>Easily create interactive graphics</p>
<pre class="r"><code>library(plotly)</code></pre>
<pre><code>## 
## Attaching package: &#39;plotly&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:ggplot2&#39;:
## 
##     last_plot</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<pre><code>## The following object is masked from &#39;package:graphics&#39;:
## 
##     layout</code></pre>
<pre class="r"><code>p &lt;- ggplot(df, aes(Coverage, Insertions, label=Accession)) +
  geom_point() +
  theme_bw()

ggplotly(p)</code></pre>
<div id="htmlwidget-1" style="width:672px;height:480px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"data":[{"x":[41,67,47,38,42,42,60,39,48,53,57,42,51,66,53,56,46,51,45,20,72,48,83,59,57,93,133,50,36,74,20,79,76,29,55,42,34,89,46,45,56,29,50,42,87,27,46,46,78,40,41,47,21,38,42,58,107,40,25,48,60,75,54,51,79,34,55,46,48,33,96,55,62,58,63,47,38,39,44,46,32,62,49,39,40,41,38,40,22,33,34,36,35,49,97,65,66,110,47,40,46,43,100,37,48,36,11,50,55,86,46,56,102,127,42,42,33,35,47,76,55,57,35,66,41,36,29,75,41,47,53,40,38,64,62,58,96,39,50,23,18,32,67,49,28,40,40,86,46,36,52,46,45,78,19,92,17,19,30,75,32,46,96,46,41,39,31,37,30,45,78,73,26,15,40,36,33,22,40,43,34,21,75,63,21,37,38,35,20,44,47,47,33,53,87,73,49,67,38,52,48,23,20,53,52,54,43,17,41,29,43,52,29,78,41,19],"y":[367,420,443,305,379,417,330,313,350,343,439,393,377,406,488,408,380,348,339,318,362,437,585,467,440,459,497,367,362,436,243,360,448,332,348,464,351,467,403,435,335,446,559,353,405,348,392,407,427,306,353,354,224,362,371,363,450,374,324,289,356,376,467,438,433,445,358,404,373,374,445,437,345,531,392,385,361,253,446,307,384,388,347,423,318,393,392,298,237,357,353,352,303,362,461,408,627,431,346,370,360,341,407,377,270,376,153,429,406,443,410,399,500,138,325,310,351,320,337,692,366,353,362,417,254,236,292,486,234,312,320,291,418,501,529,424,482,357,236,220,240,378,448,430,312,326,311,421,289,279,401,412,333,502,279,555,191,229,340,411,352,427,457,387,370,444,369,275,321,396,408,462,321,180,317,306,372,236,284,442,369,272,378,408,212,385,340,363,204,459,380,354,339,328,424,461,403,431,307,337,449,286,182,380,336,401,322,181,453,424,409,346,288,468,450,249],"text":["Coverage:  41<br />Insertions: 367<br />Accession: Aa-0","Coverage:  67<br />Insertions: 420<br />Accession: Abd-0","Coverage:  47<br />Insertions: 443<br />Accession: Ag-0","Coverage:  38<br />Insertions: 305<br />Accession: Ak-1","Coverage:  42<br />Insertions: 379<br />Accession: Alst-1","Coverage:  42<br />Insertions: 417<br />Accession: Altai-5","Coverage:  60<br />Insertions: 330<br />Accession: Amel-1","Coverage:  39<br />Insertions: 313<br />Accession: An-1","Coverage:  48<br />Insertions: 350<br />Accession: Ang-0","Coverage:  53<br />Insertions: 343<br />Accession: Anholt-1","Coverage:  57<br />Insertions: 439<br />Accession: Ann-1","Coverage:  42<br />Insertions: 393<br />Accession: Anz-0","Coverage:  51<br />Insertions: 377<br />Accession: Appt-1","Coverage:  66<br />Insertions: 406<br />Accession: Ba-1","Coverage:  53<br />Insertions: 488<br />Accession: Baa-1","Coverage:  56<br />Insertions: 408<br />Accession: Bch-1","Coverage:  46<br />Insertions: 380<br />Accession: Bd-0","Coverage:  51<br />Insertions: 348<br />Accession: Benk-1","Coverage:  45<br />Insertions: 339<br />Accession: Ber","Coverage:  20<br />Insertions: 318<br />Accession: Bg-2","Coverage:  72<br />Insertions: 362<br />Accession: Bik-1","Coverage:  48<br />Insertions: 437<br />Accession: Bl-1","Coverage:  83<br />Insertions: 585<br />Accession: Bla-1","Coverage:  59<br />Insertions: 467<br />Accession: Boot-1","Coverage:  57<br />Insertions: 440<br />Accession: Bor-1","Coverage:  93<br />Insertions: 459<br />Accession: Bor-4","Coverage: 133<br />Insertions: 497<br />Accession: Br-0","Coverage:  50<br />Insertions: 367<br />Accession: Bs-1","Coverage:  36<br />Insertions: 362<br />Accession: Bsch-0","Coverage:  74<br />Insertions: 436<br />Accession: Bu-0","Coverage:  20<br />Insertions: 243<br />Accession: Buckhorn-Pass","Coverage:  79<br />Insertions: 360<br />Accession: Ca-0","Coverage:  76<br />Insertions: 448<br />Accession: Cal-0","Coverage:  29<br />Insertions: 332<br />Accession: Cerv-1","Coverage:  55<br />Insertions: 348<br />Accession: Chat-1","Coverage:  42<br />Insertions: 464<br />Accession: Chi-0","Coverage:  34<br />Insertions: 351<br />Accession: CIBC-17","Coverage:  89<br />Insertions: 467<br />Accession: CIBC-5","Coverage:  46<br />Insertions: 403<br />Accession: Cnt-1","Coverage:  45<br />Insertions: 435<br />Accession: Co-1","Coverage:  56<br />Insertions: 335<br />Accession: Com-1","Coverage:  29<br />Insertions: 446<br />Accession: Co","Coverage:  50<br />Insertions: 559<br />Accession: Cvi-0","Coverage:  42<br />Insertions: 353<br />Accession: Da_1_-12","Coverage:  87<br />Insertions: 405<br />Accession: Db-1","Coverage:  27<br />Insertions: 348<br />Accession: Dem-4","Coverage:  46<br />Insertions: 392<br />Accession: Di-G","Coverage:  46<br />Insertions: 407<br />Accession: Dja-1","Coverage:  78<br />Insertions: 427<br />Accession: Do-0","Coverage:  40<br />Insertions: 306<br />Accession: Dr-0","Coverage:  41<br />Insertions: 353<br />Accession: Dra-0","Coverage:  47<br />Insertions: 354<br />Accession: DraII-1","Coverage:  21<br />Insertions: 224<br />Accession: DraIII-1","Coverage:  38<br />Insertions: 362<br />Accession: Duk","Coverage:  42<br />Insertions: 371<br />Accession: Durh-1","Coverage:  58<br />Insertions: 363<br />Accession: Ei-2","Coverage: 107<br />Insertions: 450<br />Accession: El-0","Coverage:  40<br />Insertions: 374<br />Accession: Ema-1","Coverage:  25<br />Insertions: 324<br />Accession: En-1","Coverage:  48<br />Insertions: 289<br />Accession: En-2","Coverage:  60<br />Insertions: 356<br />Accession: En-D","Coverage:  75<br />Insertions: 376<br />Accession: Er-0","Coverage:  54<br />Insertions: 467<br />Accession: Es-0","Coverage:  51<br />Insertions: 438<br />Accession: Est","Coverage:  79<br />Insertions: 433<br />Accession: Et-0","Coverage:  34<br />Insertions: 445<br />Accession: Etna-2","Coverage:  55<br />Insertions: 358<br />Accession: Fi-0","Coverage:  46<br />Insertions: 404<br />Accession: Fr-2","Coverage:  48<br />Insertions: 373<br />Accession: Ga-0","Coverage:  33<br />Insertions: 374<br />Accession: Ge-0","Coverage:  96<br />Insertions: 445<br />Accession: Gel-1","Coverage:  55<br />Insertions: 437<br />Accession: Gie-0","Coverage:  62<br />Insertions: 345<br />Accession: Gifu-2","Coverage:  58<br />Insertions: 531<br />Accession: Got-22","Coverage:  63<br />Insertions: 392<br />Accession: Got-7","Coverage:  47<br />Insertions: 385<br />Accession: Gr-1","Coverage:  38<br />Insertions: 361<br />Accession: Gre-0","Coverage:  39<br />Insertions: 253<br />Accession: Gu-0","Coverage:  44<br />Insertions: 446<br />Accession: Gy-0","Coverage:  46<br />Insertions: 307<br />Accession: Ha-0","Coverage:  32<br />Insertions: 384<br />Accession: Hau-0","Coverage:  62<br />Insertions: 388<br />Accession: Hey-1","Coverage:  49<br />Insertions: 347<br />Accession: Hh-0","Coverage:  39<br />Insertions: 423<br />Accession: Hn-0","Coverage:  40<br />Insertions: 318<br />Accession: Hod","Coverage:  41<br />Insertions: 393<br />Accession: HR-10","Coverage:  38<br />Insertions: 392<br />Accession: HR-5","Coverage:  40<br />Insertions: 298<br />Accession: Hs-0","Coverage:  22<br />Insertions: 237<br />Accession: HSm","Coverage:  33<br />Insertions: 357<br />Accession: In-0","Coverage:  34<br />Insertions: 353<br />Accession: Is-0","Coverage:  36<br />Insertions: 352<br />Accession: Je-0","Coverage:  35<br />Insertions: 303<br />Accession: Jl-3","Coverage:  49<br />Insertions: 362<br />Accession: Jm-0","Coverage:  97<br />Insertions: 461<br />Accession: Kar-1","Coverage:  65<br />Insertions: 408<br />Accession: Kas-1","Coverage:  66<br />Insertions: 627<br />Accession: Kas-2","Coverage: 110<br />Insertions: 431<br />Accession: Kb-0","Coverage:  47<br />Insertions: 346<br />Accession: Kelsterbach-4","Coverage:  40<br />Insertions: 370<br />Accession: Kent","Coverage:  46<br />Insertions: 360<br />Accession: Kil-0","Coverage:  43<br />Insertions: 341<br />Accession: Kin-0","Coverage: 100<br />Insertions: 407<br />Accession: Kl-5","Coverage:  37<br />Insertions: 377<br />Accession: Knox-10","Coverage:  48<br />Insertions: 270<br />Accession: Knox-18","Coverage:  36<br />Insertions: 376<br />Accession: Ko-2","Coverage:  11<br />Insertions: 153<br />Accession: Koln","Coverage:  50<br />Insertions: 429<br />Accession: Kondara","Coverage:  55<br />Insertions: 406<br />Accession: Kro-0","Coverage:  86<br />Insertions: 443<br />Accession: Krot-0","Coverage:  46<br />Insertions: 410<br />Accession: Kyoto","Coverage:  56<br />Insertions: 399<br />Accession: Kz-9","Coverage: 102<br />Insertions: 500<br />Accession: La-0","Coverage: 127<br />Insertions: 138<br />Accession: Lan-0","Coverage:  42<br />Insertions: 325<br />Accession: Le-0","Coverage:  42<br />Insertions: 310<br />Accession: Li-2:1","Coverage:  33<br />Insertions: 351<br />Accession: Lip-0","Coverage:  35<br />Insertions: 320<br />Accession: Lisse","Coverage:  47<br />Insertions: 337<br />Accession: Litva","Coverage:  76<br />Insertions: 692<br />Accession: LL-0","Coverage:  55<br />Insertions: 366<br />Accession: Lm-2","Coverage:  57<br />Insertions: 353<br />Accession: Lp2-2","Coverage:  35<br />Insertions: 362<br />Accession: Lp2-6","Coverage:  66<br />Insertions: 417<br />Accession: Mc-0","Coverage:  41<br />Insertions: 254<br />Accession: Mh-0","Coverage:  36<br />Insertions: 236<br />Accession: Mir-0","Coverage:  29<br />Insertions: 292<br />Accession: Mnz-0","Coverage:  75<br />Insertions: 486<br />Accession: Ms-0","Coverage:  41<br />Insertions: 234<br />Accession: Mv-0","Coverage:  47<br />Insertions: 312<br />Accession: Mz-0","Coverage:  53<br />Insertions: 320<br />Accession: Na-1","Coverage:  40<br />Insertions: 291<br />Accession: Nc-1","Coverage:  38<br />Insertions: 418<br />Accession: NC-6","Coverage:  64<br />Insertions: 501<br />Accession: Neo-6","Coverage:  62<br />Insertions: 529<br />Accession: NFA-10","Coverage:  58<br />Insertions: 424<br />Accession: NFA-8","Coverage:  96<br />Insertions: 482<br />Accession: Nok-3","Coverage:  39<br />Insertions: 357<br />Accession: Np-0","Coverage:  50<br />Insertions: 236<br />Accession: Nw-0","Coverage:  23<br />Insertions: 220<br />Accession: Ob-0","Coverage:  18<br />Insertions: 240<br />Accession: Old-1","Coverage:  32<br />Insertions: 378<br />Accession: Or-0","Coverage:  67<br />Insertions: 448<br />Accession: Ove-0","Coverage:  49<br />Insertions: 430<br />Accession: Per-1","Coverage:  28<br />Insertions: 312<br />Accession: PHW-2","Coverage:  40<br />Insertions: 326<br />Accession: PHW-34","Coverage:  40<br />Insertions: 311<br />Accession: Pi-0","Coverage:  86<br />Insertions: 421<br />Accession: Pla-0","Coverage:  46<br />Insertions: 289<br />Accession: Pna-17","Coverage:  36<br />Insertions: 279<br />Accession: Pog-0","Coverage:  52<br />Insertions: 401<br />Accession: Pro-0","Coverage:  46<br />Insertions: 412<br />Accession: Pt-0","Coverage:  45<br />Insertions: 333<br />Accession: Pu2-23","Coverage:  78<br />Insertions: 502<br />Accession: Pu2-7","Coverage:  19<br />Insertions: 279<br />Accession: Pu2-8","Coverage:  92<br />Insertions: 555<br />Accession: Qar-8a","Coverage:  17<br />Insertions: 191<br />Accession: Ra-0","Coverage:  19<br />Insertions: 229<br />Accession: Ragl-1","Coverage:  30<br />Insertions: 340<br />Accession: Rak-2","Coverage:  75<br />Insertions: 411<br />Accession: Rd-0","Coverage:  32<br />Insertions: 352<br />Accession: Ren-11","Coverage:  46<br />Insertions: 427<br />Accession: Ren-1","Coverage:  96<br />Insertions: 457<br />Accession: Rennes-1","Coverage:  46<br />Insertions: 387<br />Accession: Rhen-1","Coverage:  41<br />Insertions: 370<br />Accession: Rld-1","Coverage:  39<br />Insertions: 444<br />Accession: RMX-A02","Coverage:  31<br />Insertions: 369<br />Accession: RMX-A180","Coverage:  37<br />Insertions: 275<br />Accession: Rome-1","Coverage:  30<br />Insertions: 321<br />Accession: Rou-0","Coverage:  45<br />Insertions: 396<br />Accession: RRS-10","Coverage:  78<br />Insertions: 408<br />Accession: RRS-7","Coverage:  73<br />Insertions: 462<br />Accession: Rubezhnoe-1","Coverage:  26<br />Insertions: 321<br />Accession: Sap-0","Coverage:  15<br />Insertions: 180<br />Accession: Se-0","Coverage:  40<br />Insertions: 317<br />Accession: Seattle-0","Coverage:  36<br />Insertions: 306<br />Accession: Sei-0","Coverage:  33<br />Insertions: 372<br />Accession: Sf-1","Coverage:  22<br />Insertions: 236<br />Accession: Sg-1","Coverage:  40<br />Insertions: 284<br />Accession: Si-0","Coverage:  43<br />Insertions: 442<br />Accession: Sorbo","Coverage:  34<br />Insertions: 369<br />Accession: Sp-0","Coverage:  21<br />Insertions: 272<br />Accession: Sq-1","Coverage:  75<br />Insertions: 378<br />Accession: Sq-8","Coverage:  63<br />Insertions: 408<br />Accession: Stw-0","Coverage:  21<br />Insertions: 212<br />Accession: Su-0","Coverage:  37<br />Insertions: 385<br />Accession: Sus-1","Coverage:  38<br />Insertions: 340<br />Accession: Ta-0","Coverage:  35<br />Insertions: 363<br />Accession: Tamm-27","Coverage:  20<br />Insertions: 204<br />Accession: Tamm-2","Coverage:  44<br />Insertions: 459<br />Accession: Tha-1","Coverage:  47<br />Insertions: 380<br />Accession: Ting-1","Coverage:  47<br />Insertions: 354<br />Accession: Tol-0","Coverage:  33<br />Insertions: 339<br />Accession: Ts-1","Coverage:  53<br />Insertions: 328<br />Accession: Ts-5","Coverage:  87<br />Insertions: 424<br />Accession: Tscha-1","Coverage:  73<br />Insertions: 461<br />Accession: Tu-0","Coverage:  49<br />Insertions: 403<br />Accession: Tul-0","Coverage:  67<br />Insertions: 431<br />Accession: Ty-0","Coverage:  38<br />Insertions: 307<br />Accession: Uk-1","Coverage:  52<br />Insertions: 337<br />Accession: Uod-1","Coverage:  48<br />Insertions: 449<br />Accession: Uod-7","Coverage:  23<br />Insertions: 286<br />Accession: Utrecht","Coverage:  20<br />Insertions: 182<br />Accession: Van-0","Coverage:  53<br />Insertions: 380<br />Accession: Ven-1","Coverage:  52<br />Insertions: 336<br />Accession: Vind-1","Coverage:  54<br />Insertions: 401<br />Accession: Wa-1","Coverage:  43<br />Insertions: 322<br />Accession: Wc-1","Coverage:  17<br />Insertions: 181<br />Accession: Wei-0","Coverage:  41<br />Insertions: 453<br />Accession: Westkar-4","Coverage:  29<br />Insertions: 424<br />Accession: Wil-1","Coverage:  43<br />Insertions: 409<br />Accession: Wl-0","Coverage:  52<br />Insertions: 346<br />Accession: Ws-2","Coverage:  29<br />Insertions: 288<br />Accession: Wt-5","Coverage:  78<br />Insertions: 468<br />Accession: Yo-0","Coverage:  41<br />Insertions: 450<br />Accession: Zal-1","Coverage:  19<br />Insertions: 249<br />Accession: Zdr-1"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,0,0,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":26.2283105022831,"r":7.30593607305936,"b":40.1826484018265,"l":43.1050228310502},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[4.9,139.1],"tickmode":"array","ticktext":["50","100"],"tickvals":[50,100],"categoryorder":"array","categoryarray":["50","100"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":"Coverage","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[110.3,719.7],"tickmode":"array","ticktext":["200","300","400","500","600","700"],"tickvals":[200,300,400,500,600,700],"categoryorder":"array","categoryarray":["200","300","400","500","600","700"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":"Insertions","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":[{"name":"Collaborate","icon":{"width":1000,"ascent":500,"descent":-50,"path":"M487 375c7-10 9-23 5-36l-79-259c-3-12-11-23-22-31-11-8-22-12-35-12l-263 0c-15 0-29 5-43 15-13 10-23 23-28 37-5 13-5 25-1 37 0 0 0 3 1 7 1 5 1 8 1 11 0 2 0 4-1 6 0 3-1 5-1 6 1 2 2 4 3 6 1 2 2 4 4 6 2 3 4 5 5 7 5 7 9 16 13 26 4 10 7 19 9 26 0 2 0 5 0 9-1 4-1 6 0 8 0 2 2 5 4 8 3 3 5 5 5 7 4 6 8 15 12 26 4 11 7 19 7 26 1 1 0 4 0 9-1 4-1 7 0 8 1 2 3 5 6 8 4 4 6 6 6 7 4 5 8 13 13 24 4 11 7 20 7 28 1 1 0 4 0 7-1 3-1 6-1 7 0 2 1 4 3 6 1 1 3 4 5 6 2 3 3 5 5 6 1 2 3 5 4 9 2 3 3 7 5 10 1 3 2 6 4 10 2 4 4 7 6 9 2 3 4 5 7 7 3 2 7 3 11 3 3 0 8 0 13-1l0-1c7 2 12 2 14 2l218 0c14 0 25-5 32-16 8-10 10-23 6-37l-79-259c-7-22-13-37-20-43-7-7-19-10-37-10l-248 0c-5 0-9-2-11-5-2-3-2-7 0-12 4-13 18-20 41-20l264 0c5 0 10 2 16 5 5 3 8 6 10 11l85 282c2 5 2 10 2 17 7-3 13-7 17-13z m-304 0c-1-3-1-5 0-7 1-1 3-2 6-2l174 0c2 0 4 1 7 2 2 2 4 4 5 7l6 18c0 3 0 5-1 7-1 1-3 2-6 2l-173 0c-3 0-5-1-8-2-2-2-4-4-4-7z m-24-73c-1-3-1-5 0-7 2-2 3-2 6-2l174 0c2 0 5 0 7 2 3 2 4 4 5 7l6 18c1 2 0 5-1 6-1 2-3 3-5 3l-174 0c-3 0-5-1-7-3-3-1-4-4-5-6z"},"click":"function(gd) { \n        // is this being viewed in RStudio?\n        if (location.search == '?viewer_pane=1') {\n          alert('To learn about plotly for collaboration, visit:\\n https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html');\n        } else {\n          window.open('https://cpsievert.github.io/plotly_book/plot-ly-for-collaboration.html', '_blank');\n        }\n      }"}],"cloud":false},"source":"A","attrs":{"485341d0dc98":{"x":{},"y":{},"label":{},"type":"scatter"}},"cur_data":"485341d0dc98","visdat":{"485341d0dc98":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"base_url":"https://plot.ly"},"evals":["config.modeBarButtonsToAdd.0.click"],"jsHooks":[]}</script>
<pre class="r"><code># from earlier
dmrs</code></pre>
<pre><code>## # A tibble: 13,484 x 3
##    chr    start    end
##    &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;
##  1 chr1     792    911
##  2 chr1   22665  22807
##  3 chr1   41392  41445
##  4 chr1   71457  71476
##  5 chr1   76839  77252
##  6 chr1  193333 193525
##  7 chr1  255844 255893
##  8 chr1  291887 291998
##  9 chr1  355668 356037
## 10 chr1  428072 428073
## # … with 13,474 more rows</code></pre>
<pre class="r"><code>ggplot(dmrs, aes(start)) +
  geom_density() + facet_wrap(~chr)</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/more_plots-1.png" width="672" /></p>
<pre class="r"><code># cleaned up
ggplot(dmrs, aes(start/1000)) +
  geom_density(adjust = 1/10, fill = &quot;blue&quot;, alpha = 0.5) +
  facet_wrap(~chr, scales = &quot;free&quot;) + theme_classic() +
  xlab(&quot;Position (kb)&quot;) + ggtitle(&quot;C-DMR Positions&quot;) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))</code></pre>
<p><img src="../../blog/r_demo_files/figure-html/more_plots-2.png" width="672" /></p>
</div>
<div id="reporting" class="section level2">
<h2>Reporting</h2>
<p>Use <a href="http://rmarkdown.rstudio.com/">RMarkdown</a> (included in RStudio). Press Knit HTML / Knit PDF to render the output document.</p>
<div id="displaying-nice-tables" class="section level3">
<h3>Displaying nice tables</h3>
<pre class="r"><code># from before
data</code></pre>
<pre><code>## # A tibble: 500 x 5
##    position experiment_1 experiment_2 experiment_3 experiment_4
##       &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
##  1        1        0.295        0           0.691         0.224
##  2        2        0.262        0.194       0.0300        0.282
##  3        3        0.223        0.170       0.647         0.302
##  4        4        0.186        0.353       0.673         0.208
##  5        5        0.253        0.756       0.525         0.284
##  6        6        0.127        0           0.692         0.340
##  7        7        0.148        0.301       0.199         0.401
##  8        8        0.220        0.133       0.319         0.349
##  9        9        0.189        0.474       0.692         0.415
## 10       10        0.197        0.425       0.992         0.380
## # … with 490 more rows</code></pre>
<pre class="r"><code># using kable
head(data) %&gt;% knitr::kable()</code></pre>
<table>
<thead>
<tr class="header">
<th align="right">position</th>
<th align="right">experiment_1</th>
<th align="right">experiment_2</th>
<th align="right">experiment_3</th>
<th align="right">experiment_4</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">0.2947252</td>
<td align="right">0.0000000</td>
<td align="right">0.6905627</td>
<td align="right">0.2240412</td>
</tr>
<tr class="even">
<td align="right">2</td>
<td align="right">0.2624428</td>
<td align="right">0.1944636</td>
<td align="right">0.0299559</td>
<td align="right">0.2823043</td>
</tr>
<tr class="odd">
<td align="right">3</td>
<td align="right">0.2227268</td>
<td align="right">0.1696682</td>
<td align="right">0.6465708</td>
<td align="right">0.3024177</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">0.1863447</td>
<td align="right">0.3532347</td>
<td align="right">0.6733737</td>
<td align="right">0.2075265</td>
</tr>
<tr class="odd">
<td align="right">5</td>
<td align="right">0.2532761</td>
<td align="right">0.7560800</td>
<td align="right">0.5250720</td>
<td align="right">0.2844748</td>
</tr>
<tr class="even">
<td align="right">6</td>
<td align="right">0.1269957</td>
<td align="right">0.0000000</td>
<td align="right">0.6915689</td>
<td align="right">0.3399855</td>
</tr>
</tbody>
</table>
</div>
<div id="code-block-options" class="section level3">
<h3>Code block options</h3>
<p>Can choose to show everything, show only output (eg plots), show nothing, show only warnings, etc..</p>
</div>
<div id="output-options" class="section level3">
<h3>Output options</h3>
<p>Output can be rendered as html, pdf, or word. You can also choose a theme, and whether to show a table of contents. One important feature is the ability to have self-contained html documents by including <code>self_contained: true</code> in the rmarkdown header. This will let you send the html file to others and still have all the plots appear correct. This should be set as the default for html output.</p>
</div>
<div id="running-code-in-other-languages" class="section level3">
<h3>Running code in other languages</h3>
<p>It is possible to have code chunks that run in languages other than R. If you look at the first code chunk in this document, you will see that it runs a shell command (git). To find out more, read about <a href="http://yihui.name/knitr/demo/engines/">language engines in knitr</a>.</p>
<p>Here is an example running python:</p>
<pre class="python"><code>chromosomes = [&quot;chr&quot; + str(i) for i in xrange(1, 6)]
print(chromosomes)</code></pre>
<pre><code>## [&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr3&#39;, &#39;chr4&#39;, &#39;chr5&#39;]</code></pre>
<p>To use objects created by running code in other languages (for example, if we wanted to use the <code>chromosomes</code> variable created in the above python chunk), you will usually need to write the results to a file and then read into R from the file. There are complicated environment variable things you can do to <a href="https://raw.githubusercontent.com/yihui/knitr-examples/master/106-polyglot.Rmd">share objects</a> between languages, but these may not be worth the extra effort.</p>
</div>
</div>
<div id="further-reading" class="section level2">
<h2>Further reading</h2>
<p><a href="http://r4ds.had.co.nz/" class="uri">http://r4ds.had.co.nz/</a></p>
</div>
<div id="session-info" class="section level2">
<h2>Session info</h2>
<pre class="r"><code>devtools::session_info()</code></pre>
<pre><code>## ─ Session info ──────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.5.3 (2019-03-11)
##  os       macOS Mojave 10.14.4        
##  system   x86_64, darwin18.2.0        
##  ui       unknown                     
##  language (EN)                        
##  collate  en_AU.UTF-8                 
##  ctype    en_AU.UTF-8                 
##  tz       America/New_York            
##  date     2019-04-12                  
## 
## ─ Packages ──────────────────────────────────────────────────────────────
##  package      * version date       lib source        
##  assertthat     0.2.0   2017-04-11 [1] CRAN (R 3.5.0)
##  backports      1.1.3   2018-12-14 [1] CRAN (R 3.5.1)
##  blogdown       0.10    2019-01-09 [1] CRAN (R 3.5.2)
##  bookdown       0.9     2018-12-21 [1] CRAN (R 3.5.2)
##  Cairo          1.5-9   2015-09-26 [1] CRAN (R 3.5.1)
##  callr          3.1.1   2018-12-21 [1] CRAN (R 3.5.2)
##  cli            1.0.1   2018-09-25 [1] CRAN (R 3.5.1)
##  colorspace     1.4-0   2019-01-13 [1] CRAN (R 3.5.2)
##  crayon         1.3.4   2017-09-16 [1] CRAN (R 3.5.0)
##  crosstalk      1.0.0   2016-12-21 [1] CRAN (R 3.5.0)
##  data.table   * 1.12.0  2019-01-13 [1] CRAN (R 3.5.2)
##  desc           1.2.0   2018-05-01 [1] CRAN (R 3.5.1)
##  devtools       2.0.1   2018-10-26 [1] CRAN (R 3.5.1)
##  digest         0.6.18  2018-10-10 [1] CRAN (R 3.5.1)
##  dplyr        * 0.8.0.1 2019-02-15 [1] CRAN (R 3.5.2)
##  evaluate       0.13    2019-02-12 [1] CRAN (R 3.5.2)
##  fansi          0.4.0   2018-10-05 [1] CRAN (R 3.5.1)
##  fs             1.2.6   2018-08-23 [1] CRAN (R 3.5.1)
##  ggplot2      * 3.1.0   2018-10-25 [1] CRAN (R 3.5.1)
##  glue           1.3.0   2018-07-17 [1] CRAN (R 3.5.1)
##  gtable         0.2.0   2016-02-26 [1] CRAN (R 3.5.0)
##  highr          0.7     2018-06-09 [1] CRAN (R 3.5.0)
##  hms            0.4.2   2018-03-10 [1] CRAN (R 3.5.0)
##  htmltools      0.3.6   2017-04-28 [1] CRAN (R 3.5.0)
##  htmlwidgets    1.3     2018-09-30 [1] CRAN (R 3.5.1)
##  httpuv         1.4.5.1 2018-12-18 [1] CRAN (R 3.5.1)
##  httr           1.4.0   2018-12-11 [1] CRAN (R 3.5.1)
##  jsonlite       1.6     2018-12-07 [1] CRAN (R 3.5.1)
##  knitr          1.21    2018-12-10 [1] CRAN (R 3.5.1)
##  labeling       0.3     2014-08-23 [1] CRAN (R 3.5.0)
##  later          0.8.0   2019-02-11 [1] CRAN (R 3.5.2)
##  lattice        0.20-38 2018-11-04 [2] CRAN (R 3.5.3)
##  lazyeval       0.2.1   2017-10-29 [1] CRAN (R 3.5.0)
##  magrittr       1.5     2014-11-22 [1] CRAN (R 3.5.0)
##  Matrix         1.2-15  2018-11-01 [2] CRAN (R 3.5.3)
##  memoise        1.1.0   2017-04-21 [1] CRAN (R 3.5.0)
##  mime           0.6     2018-10-05 [1] CRAN (R 3.5.1)
##  munsell        0.5.0   2018-06-12 [1] CRAN (R 3.5.0)
##  pillar         1.3.1   2018-12-15 [1] CRAN (R 3.5.1)
##  pkgbuild       1.0.2   2018-10-16 [1] CRAN (R 3.5.1)
##  pkgconfig      2.0.2   2018-08-16 [1] CRAN (R 3.5.1)
##  pkgload        1.0.2   2018-10-29 [1] CRAN (R 3.5.1)
##  plotly       * 4.8.0   2018-07-20 [1] CRAN (R 3.5.1)
##  plyr           1.8.4   2016-06-08 [1] CRAN (R 3.5.0)
##  prettyunits    1.0.2   2015-07-13 [1] CRAN (R 3.5.0)
##  processx       3.2.1   2018-12-05 [1] CRAN (R 3.5.1)
##  promises       1.0.1   2018-04-13 [1] CRAN (R 3.5.0)
##  ps             1.3.0   2018-12-21 [1] CRAN (R 3.5.2)
##  purrr          0.3.0   2019-01-27 [1] CRAN (R 3.5.2)
##  R6             2.4.0   2019-02-14 [1] CRAN (R 3.5.2)
##  RColorBrewer * 1.1-2   2014-12-07 [1] CRAN (R 3.5.0)
##  Rcpp           1.0.0   2018-11-07 [1] CRAN (R 3.5.1)
##  readr        * 1.3.1   2018-12-21 [1] CRAN (R 3.5.2)
##  remotes        2.0.2   2018-10-30 [1] CRAN (R 3.5.1)
##  reticulate     1.11    2019-02-27 [1] CRAN (R 3.5.2)
##  rlang          0.3.1   2019-01-08 [1] CRAN (R 3.5.2)
##  rmarkdown      1.11    2018-12-08 [1] CRAN (R 3.5.2)
##  rprojroot      1.3-2   2018-01-03 [1] CRAN (R 3.5.1)
##  rstudioapi     0.9.0   2019-01-09 [1] CRAN (R 3.5.2)
##  scales         1.0.0   2018-08-09 [1] CRAN (R 3.5.1)
##  sessioninfo    1.1.1   2018-11-05 [1] CRAN (R 3.5.1)
##  shiny          1.2.0   2018-11-02 [1] CRAN (R 3.5.1)
##  stringi        1.3.1   2019-02-13 [1] CRAN (R 3.5.2)
##  stringr        1.4.0   2019-02-10 [1] CRAN (R 3.5.2)
##  testthat       2.0.1   2018-10-13 [1] CRAN (R 3.5.1)
##  tibble         2.0.1   2019-01-12 [1] CRAN (R 3.5.2)
##  tidyr        * 0.8.3   2019-03-01 [1] CRAN (R 3.5.2)
##  tidyselect     0.2.5   2018-10-11 [1] CRAN (R 3.5.1)
##  usethis        1.4.0   2018-08-14 [1] CRAN (R 3.5.1)
##  utf8           1.1.4   2018-05-24 [1] CRAN (R 3.5.0)
##  viridisLite    0.3.0   2018-02-01 [1] CRAN (R 3.5.0)
##  withr          2.1.2   2018-03-15 [1] CRAN (R 3.5.0)
##  xfun           0.5     2019-02-20 [1] CRAN (R 3.5.2)
##  xtable         1.8-3   2018-08-29 [1] CRAN (R 3.5.1)
##  yaml           2.2.0   2018-07-25 [1] CRAN (R 3.5.1)
## 
## [1] /usr/local/lib/R/3.5/site-library
## [2] /usr/local/Cellar/r/3.5.3/lib/R/library</code></pre>
</div>



                    </div>

                    
                        <div class="navigation">
                            <div class="row">
                                <div class="col-12 col-md-6">
                                    
                                        <div class="mx-0 mx-md-4 mt-4 text-left">
                                            <a href="../../blog/2016-07-12-smrna-analysis/">« smRNA analysis notes</a>
                                        </div>
                                    
                                </div>
                                <div class="col-12 col-md-6">
                                    
                                        <div class="mx-0 mx-md-4 mt-4 text-right">
                                            <a href="../../blog/2017-03-03-installing-magic/">Installing Magic »</a>
                                        </div>
                                    
                                </div>
                            </div>
                        </div>
                    
                </div>
            </div>
        </div>
    </div>

    <section id="comments">

      <div class="py-3 content">
            <div class="container">
                  <div class="row justify-content-center">
                        <div class="col-sm-12 col-md-10">
                              <div class="comments">
                                    <div id="disqus_thread"></div>
                              </div>
                        </div>
                  </div>
            </div>
      </div>

      <script type="text/javascript">
            (function () {
                  
                  
                  if (window.location.hostname == "localhost")
                        return;

                  var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                  var disqus_shortname = '';
                  dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                  (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
      </script>
      <noscript>
            Please enable JavaScript to view the
            <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
      </noscript>
</section>
    <div class="my-4 footer">
    <div class="container">
        <div class="row justify-content-center">
            <div class="col-sm-12 col-md-5">
                
            </div>
            <div class="col-sm-12 col-md-5">
                <div class="mx-0 mx-md-4 text-right">
                    
                        <a href="https://github.com/timoast" target="_blank">
                            <img class="icon" src="../../img/github.svg" alt="github" />
                        </a>
                    

                    

                    
                        <a href="https://twitter.com/timoast" target="_blank">
                            <img class="icon" src="../../img/twitter.svg" alt="twitter" />
                        </a>
                    

                    

                    

                    

                    

                    
                        <a href="../../index.xml" class="mr-0">
                            <img class="icon" src="../../img/rss.svg" alt="rss" />
                        </a>
                    

                    
                </div>
            </div>
        </div>
    </div>
</div>



    

    
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.13.1/highlight.min.js" integrity="sha256-iq71rXEe/fvjCUP9AfLY0cKudQuKAQywiUpXkRFSkLc=" crossorigin="anonymous"></script>
        
        <script>
            window.addEventListener('load', function() {
                hljs.initHighlighting();
            }, true);
        </script>
    

    

    
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-131769435-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>

    
        
<script src="../../js/math-code.js"></script>
<script async src="//cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
</body>

</html>
