<!doctype html>
<html lang="en-us">

<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title> Useful Linux/Unix commands - Tim Stuart </title>
    <meta name="HandheldFriendly" content="True">
    <meta name="MobileOptimized" content="320">
    <meta name="referrer" content="no-referrer">
    <meta name="description" content="Tim Stuart" />
    <meta property="og:site_name" content="Tim Stuart" />
    <meta property="og:locale" content="nn_NO" />
    <meta property="og:type" content="article" />
    <meta property="og:url" content="/blog/2014-08-24-usefulcommands/" />
    <meta property="og:title" content="Useful Linux/Unix commands" />
    <meta property="og:image" content="/" />
    <meta property="og:description" content="Tim Stuart" />

    
        <meta property="twitter:site" content="@timoast">
    

    <meta property="twitter:title" content="Useful Linux/Unix commands" />
    <meta property="twitter:image" content="/" />
    <meta property="twitter:card" content="summary" />
    <meta property="twitter:description" content="Tim Stuart" />

    <link rel="canonical" href="../../blog/2014-08-24-usefulcommands/">

    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha256-eSi1q2PG6J7g7ib17yAaWMcrr5GrtohYChqibrV7PBE=" crossorigin="anonymous" />

    <link rel="stylesheet" href="../../css/custom.css" />

    
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.13.1/styles/tomorrow.min.css" integrity="sha256-0QU8ry64q+N6YBIEF/6XF6vUeF15gbNO4tLS6ikk0FI=" crossorigin="anonymous" />
    

    

    <link rel="shortcut icon"
        href="../../img/favicon.ico">

    
</head>

<body>
    
    <div class="my-4 my-md-5 header">
    <div class="container">
        <div class="row">
            <div class="col-auto offset-md-1 d-none d-md-block">
                
            </div>
            <div class="col-auto align-self-center mr-auto">
                <a href="../../">
                    <h1 class="name">Tim Stuart</h1>
                </a>

                <ul class="nav nav-primary">
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-about" href="../../">
                                
                                About
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-research" href="../../research/">
                                
                                Research
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-publications" href="../../publications/">
                                
                                Publications
                            </a>
                        </li>
                    
                        <li class="nav-item">
                            <a class="text-uppercase nav-link text-blog" href="../../blog/">
                                
                                Blog
                            </a>
                        </li>
                    

                    
                </ul>

            </div>
        </div>
    </div>
</div>


    <div class="content">
        <div class="container">
            <div class="row justify-content-center">
                <div class="col-sm-12 col-md-10">
                    <h1 class="mx-0 mx-md-4 blog-post-title">Useful Linux/Unix commands</h1>

                    <div class="mb-md-4 meta">
                        
                            
                        

                        <span class="date middot" title='Sun Aug 24 2014 00:00:00 UTC'>
                            2014-08-24
                        </span>

                        <span class="reading-time middot">
                            6 minute read
                        </span>

                        <div class="d-none d-md-inline tags">
                            <ul class="list-unstyled d-inline">
                                
                            </ul>
                        </div>

                        <div class="d-none d-md-inline tags">
                            <ul class="list-unstyled d-inline">
                                
                                
                            </ul>
                        </div>
                    </div>

                    <div class="markdown">
                        
    

<h2 id="go-back-to-last-directory">go back to last directory</h2>

<pre><code>cd -
</code></pre>

<h2 id="do-something-in-a-range">Do something in a range</h2>

<p>Move files</p>

<pre><code class="language-bash">for i in $(seq 10); do mv chr$i ../genomes/; done
</code></pre>

<pre><code class="language-bash">for directory in ./*; do
    if [ -d &quot;$directory&quot; ]; then
        cd $directory
        for myfile in $(ls -d *.sra);do
            mv $myfile /home/tstuart/working_data/1000genomes/$myfile
        done
        cd ..
    fi
done
</code></pre>

<!--break-->

<h2 id="split-fasta-file-into-a-new-file-for-each-entry">Split fasta file into a new file for each entry</h2>

<pre><code>awk '/&gt;/{x=&quot;file_&quot;++i;}{print &gt; x&quot;.txt&quot;;}' input
</code></pre>

<h2 id="split-coordinate-file-into-different-files-for-each-chromosome">Split coordinate file into different files for each chromosome</h2>

<pre><code>awk  '{print &gt; $1&quot;.txt&quot;}' input
</code></pre>

<h2 id="take-command-line-arguments">Take command-line arguments</h2>

<p>Required flags are followed by <code>:</code></p>

<pre><code class="language-bash">index=  proc=  path=  

while getopts x:pA opt; do
  case $opt in
  x)
      index=$OPTARG
      ;;
  p)
      proc=$OPTARG
      ;;
  A)
      path=${OPTARG%/}
      ;;
  esac
done
shift $((OPTIND - 1))
</code></pre>

<h2 id="strip-trailing-slash-from-string">Strip trailing slash from string</h2>

<pre><code class="language-bash">path=${OPTARG%/}
</code></pre>

<h2 id="gnu-screen">GNU Screen</h2>

<p>Start screen: <code>screen -S [screen name]</code><br />
List screens: <code>screen -list</code><br />
Detach: <code>screen -d</code><br />
Attach: <code>screen -r</code><br />
Close screen: <code>ctr-a-d</code><br />
Kill screen: <code>ctr-a :quit</code></p>

<h2 id="human-readable-path">Human-readable path</h2>

<pre><code class="language-bash">echo -e ${PATH//:/'\n'}
</code></pre>

<h2 id="searching">Searching</h2>

<pre><code class="language-bash">grep 'text' filename
grep 'text' file1 file2 file3
grep 'text1 text2' filename
grep --color 'text' filename
</code></pre>

<p>Search all files in a directory, show output in less</p>

<pre><code class="language-bash">grep -r 'text' /home/usr/ | less
</code></pre>

<p>Search for multiple strings</p>

<pre><code class="language-bash">egrep '(AT4G40030)|(AT4G40040)|(AT5G10980)' * &gt; h3.3
</code></pre>

<h2 id="pipe-output-into-file">Pipe output into file</h2>

<h4 id="overwrite-contents-of-file">Overwrite contents of file</h4>

<pre><code class="language-bash">grep &quot;text&quot; filename.txt &gt; output.txt
</code></pre>

<h4 id="append-to-file">Append to file</h4>

<pre><code class="language-bash">grep &quot;text&quot; filename.txt &gt;&gt; output.txt
</code></pre>

<h2 id="counting">Counting</h2>

<p>Count lines in chr1 file</p>

<pre><code class="language-bash">wc -l chr1
</code></pre>

<p>Count characters in chr1 file</p>

<pre><code class="language-bash">wc -c chr1
</code></pre>

<h2 id="list">List</h2>

<p>List and sort</p>

<pre><code class="language-bash">ls -ls
</code></pre>

<p>List and sort by size</p>

<pre><code class="language-bash">ls -ls -S
</code></pre>

<p>Count number of directories in current directory</p>

<pre><code class="language-bash">ls -l | grep ^d | wc -l
</code></pre>

<h2 id="sorting">Sorting</h2>

<p>Sort file by numerical order of first column, save as sorted_list.txt</p>

<pre><code class="language-bash">sort -nk 1 list.txt &gt; sorted_list.txt
</code></pre>

<p>Sort file by order of first column, then numerical order of second column, save as sorted_list.txt</p>

<pre><code class="language-bash">sort -k1,1 -nk2,2 list.txt &gt; sorted_list.txt
</code></pre>

<p>Sort descending</p>

<pre><code class="language-bash">sort -rn -k3 file.txt &gt; sorted.txt
</code></pre>

<p>overwrite</p>

<pre><code class="language-bash">sort -rn -k3 -o file.txt file.txt
</code></pre>

<h2 id="rows">Rows</h2>

<p>Delete first line</p>

<pre><code class="language-bash">sed 1d file.txt &gt; headerless_file.txt
</code></pre>

<p>Delete lines 1-3 inclusive</p>

<pre><code class="language-bash">sed 1,3d file.txt &gt; newfile.txt
</code></pre>

<p>Delete lines containing string</p>

<pre><code class="language-bash">sed '/ATC/d' file.txt &gt; file_mod.txt
</code></pre>

<p>Overwrite file</p>

<pre><code class="language-bash">sed -i.bak 1d file.txt
</code></pre>

<h2 id="columns">Columns</h2>

<p>Write columns to new file. Can also reorder columns.</p>

<pre><code class="language-bash">awk 'BEGIN {FS=OFS=&quot;\t&quot;} {print $2,$4,$5,$6,$7,$8,$1,$3}' file.txt &gt; outfile.txt
awk 'BEGIN {FS=OFS=&quot;\t&quot;} {print $7,$1,$8,$2,$3,$4,$5,$6}' cmt2_targets.tsv &gt; cmt2_targets_ordered.tsv
</code></pre>

<p>Add column</p>

<pre><code class="language-bash">awk '{print $0, &quot;cmt2&quot;}' cmt2_targets_ordered.tsv &gt; newfile.tsv
</code></pre>

<p>Remove first column from all files</p>

<pre><code class="language-bash">for myfile in $(ls);do
    awk 'BEGIN {FS=OFS=&quot;\t&quot;} {$1=&quot;&quot;;sub(&quot;\t&quot;,&quot;&quot;)}1' $myfile &gt; mod_$myfile
done
</code></pre>

<p>Search only in one column</p>

<pre><code class="language-bash">awk '{if ($1 == 1) print $2}' p1 &gt; chr1_p1
</code></pre>

<p>Add string prefix to a column, eg. add &lsquo;chr&rsquo; to the first column</p>

<pre><code class="language-bash">awk 'BEGIN {FS=OFS=&quot;\t&quot;} {$1=&quot;chr&quot;$1; print $0}' input_file.txt &gt; output_file.txt
</code></pre>

<h2 id="joining-files">Joining files</h2>

<p>Join files with each starting on a new line</p>

<pre><code class="language-bash">for filename in *.fa; do
    cat &quot;${filename}&quot;
    echo
done &gt; output.fa
</code></pre>

<p>Merge files of the same format</p>

<pre><code class="language-bash">cat p1_* | sort -nk1,1 -k2,2 | uniq &gt; p1
</code></pre>

<h2 id="comparing-files">Comparing Files</h2>

<p>Files should first be sorted</p>

<p>Find lines that are common or different between two files</p>

<pre><code class="language-bash">comm -i file1 file2 &gt; output.txt
</code></pre>

<ul>
<li><code>-i</code> is case-insensitive<br /></li>
<li>outputs three columns. First is lines only in file1, second is lines only in file2, third is lines common in both.</li>
</ul>

<h2 id="compressing-files">Compressing files</h2>

<p>Compress recursively and store all files in a single compressed folder</p>

<pre><code class="language-bash">tar cvfz slam.tgz slam/
</code></pre>

<p>Compress recursively</p>

<pre><code class="language-bash">gzip -r directory/
</code></pre>

<p>Decompress files</p>

<pre><code class="language-bash">tar -x -f file.tar
tar -x -z -f file.tgz
gunzip file.gz
</code></pre>

<h2 id="download-data-from-sra">Download data from SRA</h2>

<h3 id="using-wget">Using wget</h3>

<pre><code>wget -r --no-parent --reject &quot;index.html*&quot; ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByStudy/sra/SRP/SRP005/SRP005399/
</code></pre>

<p>Alternatively, you can download a list of accessions from the SRA website and use that file to call <code>wget</code> for all the accessions.</p>

<pre><code class="language-python">from subprocess import call


with open('SraAccList.txt', 'r') as accessions:
    for row in accessions:
        acc = row.strip('\n')
        prefix = acc[:6]
        print &quot;Downloading {a}&quot;.format(a=acc)
        call([&quot;wget&quot;,
              &quot;-r&quot;,
              &quot;--no-parent&quot;,
              &quot;ftp://ftp-trace.ncbi.nih.gov/sra/sra-instant/reads/ByRun/sra/SRR/{pre}/{acc}/*&quot;.format(pre=prefix, acc=acc)])

</code></pre>

<h4 id="split-into-fastq-files">Split into fastq files</h4>

<pre><code class="language-bash">fastq-dump --split-3 ./SRP005399/*
</code></pre>

<p><code>--split-3</code> causes PE files to be split into <code>*_1.fastq</code> and <code>*_2.fastq</code></p>

<p>bash loop</p>

<pre><code class="language-bash">for directory in ./*; do
    if [ -d &quot;$directory&quot; ]; then
        cd $directory
        for myfile in $(ls -d *.sra);do
            fastq-dump --split-3 -v $myfile
        done
        cd ..
    fi
done
</code></pre>

<p>Python function</p>

<pre><code class="language-python">import os
from subprocess import call


def fastqSplit():
    for filename in os.listdir('.'):
        if filename.endswith('.sra'):
            print 'processing {n}'.format(n=filename)
            call(['fastq-dump', '--split-3', '-v', filename])
        else:
            pass
</code></pre>

<h3 id="using-sra-toolkit">Using SRA toolkit</h3>

<p>SRA toolkit doesn&rsquo;t seem to work well for downloading bulk data. <code>wget</code> is a much better option as it allows download of whole folders, gives more descriptive output as it goes. SRA often fails and sometimes gives no error. Download with <code>wget</code> and use <code>fastq-split</code> to get fastq files.</p>

<pre><code class="language-bash">fastq-dump SRR534224 &amp;
</code></pre>

<h2 id="rna-seq">RNA-seq</h2>

<h3 id="mapping">Mapping</h3>

<pre><code>tophat -p8 -G /home/lister/working_data/data/genomes/annotations/tair10/TAIR10_gene_TE_illumina.gtf --transcriptome-index=transcript_index/tair10 /home/lister/working_data/data/genomes/bowtie2_indexes/tair9 SRR501604.fastq,SRR501605.fastq
</code></pre>

<ul>
<li><code>--transcriptome-index=transcript_index/tair10</code> creates transcriptome index file saved in transcript_index/ with name <code>tair10.*</code>. This can be reused if mapping to the same transcriptome (ie. gff file)<br /></li>
<li><code>-p8</code> uses 8 cores<br /></li>
<li><code>-G</code> specifies gff file. Optional<br /></li>
<li>If using PE reads check manual<br /></li>
</ul>

<h2 id="bedtools">Bedtools</h2>

<p><a href="http://quinlanlab.org/tutorials/cshl2014/bedtools.html">Good tutorial from the Quinlan lab</a></p>

<p>Need to order files so that chromosome, start, stop and first three columns (BED format). Also need to remove header. Can do these steps with <code>awk</code> and <code>sed 1d</code>.</p>

<h3 id="merge">Merge</h3>

<pre><code class="language-bash">bedtools merge -i dmrs_ddc.tsv -d 100 &gt; dmrs_ddc_merged.tsv
</code></pre>

<ul>
<li><code>-d</code> is distance between coordinates that will still be merged<br /></li>
<li><code>-i</code> is input file<br /></li>
</ul>

<h3 id="intersect">Intersect</h3>

<pre><code class="language-bash">bedtools intersect -a tair10_tes.txt -b dmrs_ddc_merged.tsv -wa -f 0.50 &gt; ddc_targets.tsv
</code></pre>

<ul>
<li><code>-a</code> is file a<br /></li>
<li><code>-b</code> is file b<br /></li>
<li><code>-wa</code> is write a<br /></li>
<li><code>-f</code> is minimum overlap percentage<br /></li>
</ul>

<h2 id="bowtie">Bowtie</h2>

<h3 id="mapping-pe-data">Mapping PE data</h3>

<pre><code>bowtie2 -p8 --local --fr -q -R 5 -N 1 -x [path/to/bowtie2/index] -X 1000 -1 [mate1.fq] -2 [mate2.fq] -S [alignment.sam] --un-conc ./discordant/
</code></pre>

<ul>
<li><code>-p</code> is number of processors<br /></li>
<li><code>--local</code> allows soft-clipping of reads to improve alignment<br /></li>
<li><code>-q</code> specifies that reads are in fastq format<br /></li>
<li><code>-x</code> gives path to index<br /></li>
<li><code>-S</code> gives name of output file<br /></li>
<li><code>-R</code> is number of times bowtie will try to &rsquo;re-seed&rsquo; repetitive seeds. Default 2<br /></li>
<li><code>-N</code> is number of mismatches allowed in seed. 0 or 1.<br /></li>
<li><code>--no-mixed</code> tells bowtie to find alignments only when both pairs can be aligned.<br /></li>
<li><code>--fr</code> means mate pairs are ordered in forward then reverse orientation. Can do <code>--ff</code>, <code>--rf</code>.<br /></li>
<li><code>-X</code> specifies maximum insert size. Default 500.<br /></li>
<li><code>-I</code> specifies minimum insert size. Default 0 (no minimum).<br /></li>
<li><code>--un-conc</code> specifies path and file to write discordant alignments to. Note that these are just the <code>fastq</code> reads, not alignments.</li>
</ul>

<h2 id="samtools">Samtools</h2>

<p>samtools can read from a stream, so can pipe output in from other tools (eg bowtie to get <code>.bam</code> output) or other samtools commands.</p>

<h3 id="convert-from-sam-to-bam">Convert from <code>sam</code> to <code>bam</code></h3>

<pre><code>samtools view -bS file.sam &gt; file.bam
</code></pre>

<h3 id="sort-bamfile">Sort bamfile</h3>

<pre><code>samtools sort file.bam sorted
</code></pre>

<ul>
<li>outputs <code>sorted.bam</code> file</li>
</ul>

<h3 id="samblaster">Samblaster</h3>

<p>Extracts reads from <code>.sam</code> alignment files</p>

<h4 id="get-discordant-reads">Get discordant reads</h4>

<pre><code>bowtie2 [options] | samblaster -e -d file.sam | samtools view -bS - &gt; file.bam
</code></pre>

<ul>
<li>Print output from bowtie to stderr to pipe directly into samblaster</li>
<li>Saves having to search through sam or bam file later to extract discordant reads</li>
<li>Pipe output directly into <code>samtools view</code> to save as <code>.bam</code> file (these won&rsquo;t be the reads in the samblaster output)</li>
</ul>



                    </div>

                    
                        <div class="navigation">
                            <div class="row">
                                <div class="col-12 col-md-6">
                                    
                                </div>
                                <div class="col-12 col-md-6">
                                    
                                        <div class="mx-0 mx-md-4 mt-4 text-right">
                                            <a href="../../blog/2014-08-25-usefulpython/">Useful python code »</a>
                                        </div>
                                    
                                </div>
                            </div>
                        </div>
                    
                </div>
            </div>
        </div>
    </div>

    <section id="comments">

      <div class="py-3 content">
            <div class="container">
                  <div class="row justify-content-center">
                        <div class="col-sm-12 col-md-10">
                              <div class="comments">
                                    <div id="disqus_thread"></div>
                              </div>
                        </div>
                  </div>
            </div>
      </div>

      <script type="text/javascript">
            (function () {
                  
                  
                  if (window.location.hostname == "localhost")
                        return;

                  var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                  var disqus_shortname = '';
                  dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                  (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
            })();
      </script>
      <noscript>
            Please enable JavaScript to view the
            <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
      </noscript>
</section>
    <div class="my-4 footer">
    <div class="container">
        <div class="row justify-content-center">
            <div class="col-sm-12 col-md-5">
                
            </div>
            <div class="col-sm-12 col-md-5">
                <div class="mx-0 mx-md-4 text-right">
                    
                        <a href="https://github.com/timoast" target="_blank">
                            <img class="icon" src="../../img/github.svg" alt="github" />
                        </a>
                    

                    

                    
                        <a href="https://twitter.com/timoast" target="_blank">
                            <img class="icon" src="../../img/twitter.svg" alt="twitter" />
                        </a>
                    

                    

                    

                    

                    

                    
                        <a href="../../index.xml" class="mr-0">
                            <img class="icon" src="../../img/rss.svg" alt="rss" />
                        </a>
                    

                    
                </div>
            </div>
        </div>
    </div>
</div>



    

    
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.13.1/highlight.min.js" integrity="sha256-iq71rXEe/fvjCUP9AfLY0cKudQuKAQywiUpXkRFSkLc=" crossorigin="anonymous"></script>
        
        <script>
            window.addEventListener('load', function() {
                hljs.initHighlighting();
            }, true);
        </script>
    

    

    
    
        
<script src="../../js/math-code.js"></script>
<script async src="//cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML"></script>


    
</body>

</html>
